dts/bindings/mipi-dsi/mipi-dsi-host.yaml

```diff
@@ -15,3 +15,9 @@ properties:
   "#size-cells":
     required: true
     const: 0
+
+  phy-clock:
+    type: int
+    description:
+      MIPI PHY clock frequency. Should be set to ensure clock frequency is at
+      least (pixel clock * bits per output pixel) / number of mipi data lanes
```

Comment:
we should actually align with upstream Linux bindings here,

see https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/display/dsi-controller.yaml

Is this property really needed?

- Clock frequency can likely be obtained from one of the `clocks = <>;`. See for example ST (I think you guys actually share the same MIPI-DSI IP from Synopsis):
```
clocks = <&rcc DSI_K>, <&clk_hse>, <&rcc DSI_PX>;
clock-names = "pclk", "ref", "px_clk";
                             ^~~~~~~~
```

- Bits per pixel depends on how you configure color
- Data lanes is something known from DT

Reply:
My concern with calculating this using a `clocks` property is that there won't always be a pixel clock present. For displays operating in command mode, the LCDIF is not required, as the display controller has internal RAM that is used to keep the display refreshed. In that case, the `phy-clock` minimum is still based off the resolution of the display, but the user can increase it in order to yield better transfer performance.

Linux seems to handle this when a display controller [attaches to a MIPI host via `mipi_dsi_attach`](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/panel/panel-novatek-nt35560.c#L484). As part of Linux's MIPI DSI device struct, the panel can set a minimum and maximum frequency: https://elixir.bootlin.com/linux/latest/source/include/drm/drm_mipi_dsi.h#L193. 

One option we have is to add to these parameters to the `mipi_dsi_device` struct in Zephyr. This way, display controllers can specify the clock frequency they require. However, at that point though, why use the C structure over devicetree? Both options are effectively hard coding the PHY clock frequency, but the devicetree better exposes this property to the user, in case they want to configure it for better performance.

Reply:
So if I understand correctly, in Linux this is passed by the display to the host, right? Then it's probably a display setting (which may be in DT, or, computed in the driver based on some DT info) passed to the DSI host.

Reply:
Yes, I think that's correct.

I've take more of a look at how Linux handles displays, specifically the RM68200 present in Zephyr and Linux.

There are three use cases I know of that I'd like to support:
- MIPI displays using command mode ([RM67162](https://www.nxp.com/design/development-boards/i-mx-evaluation-and-development-boards/1-2-wearable-display-g1120b0mipi:G1120B0MIPI))
- MIPI displays using video mode ([RM68200](https://www.nxp.com/design/development-boards/i-mx-evaluation-and-development-boards/5-5-lcd-panel:RK055HDMIPI4MA0))
- Displays receiving raw RGB data from the LCD interface([RK043FN66HS-CTG](https://www.nxp.com/design/development-boards/i-mx-evaluation-and-development-boards/4-3-lcd-panel:RK043FN66HS-CTG))

For a display in video mode (like the RM68200) the chain would look like the following:
FAMEBUFFER->LCDIF->MIPI-DSI->DISPLAY
For a display in command mode, (like the RM67162) the chain would look like so:
FRAMEBUFFER->MIPI-DSI->DISPLAY
There are also devices using the LCDIF output directly:
FRAMEBUFFER->LCDIF->DISPLAY

From what I can tell, [drm bridges](https://dri.freedesktop.org/docs/drm/gpu/drm-kms-helpers.html#bridges) are used to build a chain of devices involved in outputting data to the display.

In Linux, it seems that the DRM subsystem will then initialize each device in the chain, and the device has the option to provide a `mode_set` callback, which the DRM subsystem will call during init. This callback lets each device in the chain set itself up to support the clock frequency requested by the display.

For example, take a look at the radium RM68200 Linux driver (this is the same panel supported on the RT1170):
https://github.com/torvalds/linux/blob/master/drivers/gpu/drm/panel/panel-raydium-rm68200.c#L85

Clocks are given in kHZ, so this 54MHz clock would yield a roughly 51Hz framerate for the display.

This display is present on an [STM32MP157C-EV1 EVK](https://github.com/torvalds/linux/blob/master/arch/arm/boot/dts/stm32mp157c-ev1.dts)

The STM32 MIPI driver validates that the requested display clock is possible here: https://github.com/torvalds/linux/blob/master/drivers/gpu/drm/stm/dw_mipi_dsi-stm.c#L328

and the STM32 LTDC driver sets the pixel clock here: https://github.com/torvalds/linux/blob/master/drivers/gpu/drm/stm/ltdc.c#L871

To summarize, I think that the DRM subsystem will initialize each device in the chain based off the clock frequency the display expects.

If we want to align with Linux here, the devicetree might look something like this for an RM68200:

```
&mipi_dsi {
	ports {
		port@0 {
			reg = <0>;
			mipi_dsi_in: endpoint {
				remote-endpoint = <&lcdif_output>;
			};
		};
		
		port@1 {
			reg = <1>;
			mipi_dsi_out: endpoint {
				remote-endpoint = <&mipi_dsi_panel_in>;
			};
		};
	};

	status = "okay";
	dpi-color-coding = "24-bit";
	dpi-pixel-packet = "24-bit";
	dpi-video-mode = "burst";
	dpi-bllp-mode = "low-power";
	autoinsert-eotp;
	
	rm68200: rm68200@0 {
		compatible = "radium,rm68200";
		reg = <0>;
		hactive = <720>;
		vactive = <1280>;
		hsync-len = <8>;
		hfront-porch = <32>;
		hback-porch = <32>;
		vsync-len = <2>;
		vfront-porch = <16>;
		vback-porch = <14>;
		de-active = <1>;
		pixelclk-active = <1>;
		pixel-format = <MIPI_DSI_PIXFMT_RGB565>;
		backlight-gpios = <&gpio9 29 GPIO_ACTIVE_HIGH>;
		/*
		 * Pixel clock is given by the following formula:
		 * (height + vsync + vfp + vbp) *
		 * (width + hsync + hfp + hbp) * frame rate
		 */
		clock-frequency = <62346240>;
		port {
			mipi_dsi_panel_in: endpoint {
				remote-endpoint = <&mipi_dsi_out>;
			};
		};
	};
};

&lcdif {
	status = "okay";
	port {
		lcdif_output: endpoint@0 {
			reg = <0>;
			remote-endpoint = <&mipi_dsi_in>;
		};
	};
};
```

My concern here is that this approach leaves several peripherals without key initialization information. In particular, the LCDIF needs to know the pixel format, as well as panel size. This issue is handled with the MIPI layer via `mipi_dsi_attach`. But for the LCDIF layer, I don't see how we create this link without some sort of `lcdif_attach` function that can propagate this information from the display.